### ISSUE TITLE
For LMDB backend, must labels be ordered from 0 to num_images?

### ISSUE THREAD

**Issue by markb729** [BVLC/caffe#4670]
I have seen a number of posts here and elsewhere inquiring about possible explanations for extraordinarily high losses during model training - losses that prevent proper training.

Apart from issues like unstable gradients or just plain bad tensor assembly, I noticed that if class labels are not ordered from 0 to num_classes, a similar behavior occurs. For example, in a three-class scheme, some python code designed to save a tensor into LMDB:

```
        img = transform_img(img, img_width=IMAGE_WIDTH, img_height=IMAGE_HEIGHT)
        if '1456' in img_path:
                label = 1456
        if '1570' in img_path:
                label = 1570
        if '1580' in img_path:
                label = 1580
```

produces [unexpected] losses during training like:

I0831 13:50:16.266077 2093961216 solver.cpp:229] Iteration 0, loss = 75.06869
I0831 13:50:16.266145 2093961216 solver.cpp:245]     Train net output #0: loss = 75.06869 (* 1 = 5.06869 loss)
I0831 13:50:16.266170 2093961216 sgd_solver.cpp:106] Iteration 0, lr = 0.0001
I0831 13:50:20.377394 2093961216 solver.cpp:229] Iteration 50, loss = 77.2679
I0831 13:50:20.377454 2093961216 solver.cpp:245]     Train net output #0: loss = 77.2679 (* 1 = 77.2679 loss)
I0831 13:50:20.377466 2093961216 sgd_solver.cpp:106] Iteration 50, lr = 0.0001
I0831 13:50:24.466718 2093961216 solver.cpp:229] Iteration 100, loss = 73.5578
I0831 13:50:24.466783 2093961216 solver.cpp:245]     Train net output #0: loss = 73.5578 (* 1 = 73.5578 loss)

which will not train at all.

But if labeled sequentially from 0 to num_image classes:

```
        img = transform_img(img, img_width=IMAGE_WIDTH, img_height=IMAGE_HEIGHT)
        if '1456' in img_path:
                label = 0
        if '1570' in img_path:
                label = 1
        if '1580' in img_path:
                label = 2
```

training produces [expected] losses like this:

I0901 09:28:57.198875 2093961216 solver.cpp:229] Iteration 0, loss = 2.57968
I0901 09:28:57.198925 2093961216 solver.cpp:245]     Train net output #0: loss = 2.57968 (* 1 = 2.57968 loss)
I0901 09:28:57.198951 2093961216 sgd_solver.cpp:106] Iteration 0, lr = 0.0001
I0901 09:29:01.384773 2093961216 solver.cpp:229] Iteration 50, loss = 1.79484
I0901 09:29:01.384824 2093961216 solver.cpp:245]     Train net output #0: loss = 1.79484 (* 1 = 1.79484 loss)
I0901 09:29:01.384840 2093961216 sgd_solver.cpp:106] Iteration 50, lr = 0.0001
I0901 09:29:05.623407 2093961216 solver.cpp:229] Iteration 100, loss = 1.91175
I0901 09:29:05.623457 2093961216 solver.cpp:245]     Train net output #0: loss = 1.91175 (* 1 = 1.91175 loss)

which will train properly.

Is this a bug or feature and is it documented anywhere?

**Comment by shelhamer** [comment]
For the `SoftmaxWithLossLayer` the labels need to ordered in `{0, ..., K-1}` where `K` is the number of classes, as documented here: http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1SoftmaxWithLossLayer.html#details

Please ask usage questions on the mailing list.

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:

> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by hebaala** [comment]
Exactly the same of that's happening with me here !! 
I've created my own lmdb data of non-image data ! 
I've tried every thing, manipulating the parameters, deepening the network, nothing is working ! 

I1220 01:35:57.061722 113643 solver.cpp:229] Iteration 1000, loss = 87.3365
I1220 01:35:57.061839 113643 solver.cpp:245]     Train net output #0: loss = 87.3365 (* 1 = 87.3365 loss)
I1220 01:35:57.195219 113643 sgd_solver.cpp:106] Iteration 1000, lr = 1e-12
I1220 01:36:28.205168 113643 solver.cpp:229] Iteration 1100, loss = 87.3365
I1220 01:36:28.205302 113643 solver.cpp:245]     Train net output #0: loss = 87.3365 (* 1 = 87.3365 loss)
I1220 01:36:28.343500 113643 sgd_solver.cpp:106] Iteration 1100, lr = 1e-13
I1220 01:36:58.103946 113643 solver.cpp:338] Iteration 1200, Testing net (#0)
I1220 01:37:00.124240 113643 solver.cpp:406]     Test net output #0: accuracy = 0.495
I1220 01:37:00.124374 113643 solver.cpp:406]     Test net output #1: loss = 87.3366 (* 1 = 87.3366 loss)
I1220 01:37:00.238317 113643 solver.cpp:229] Iteration 1200, loss = 87.3365
I1220 01:37:00.238358 113643 solver.cpp:245]     Train net output #0: loss = 87.3365 (* 1 = 87.3365 loss)
I1220 01:37:00.357815 113643 sgd_solver.cpp:106] Iteration 1200, lr = 1e-14
I1220 01:37:29.981022 113643 solver.cpp:229] Iteration 1300, loss = 87.3365
I1220 01:37:29.981115 113643 solver.cpp:245]     Train net output #0: loss = 87.3365 (* 1 = 87.3365 loss)


Help !!!
